In [113]:
import pandas as pd
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [50]:
# recall is about real positives (higher recall in high risk)
# precision is about positive predictions, so we want to limit the false positives as much as possible 
# set the most common class as the baseline

## 2. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |



- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?

In [69]:
# cat = positive class
# dog - negative class
# False positive: we predicted a cat, but its actually a dog
# False negative: We predicted a dog, but its actually a cat

#true positive is predicting its a cat, and its a cat
tp = 34

#true negative is predicting its a dog, and its a dog
tn = 46

#false positive is predicting its a cat, but its a dog
fp = 7

#false negative is predicting its a dog, but its a cat
fn = 13

In [70]:
print("True Positives", tp)
print("False Positives", fp)
print("False Negatives", fn)
print("True Negatives", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)

precision = tp / (tp + fp)

recall = tp / (tp + fn)

print("Accuracy is", accuracy)
print("Recall is", round(recall,2))
print("Precision is", round(precision,2))

True Positives 34
False Positives 7
False Negatives 13
True Negatives 46
-------------
Accuracy is 0.8
Recall is 0.72
Precision is 0.83


## 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

 - An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
 - Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


In [75]:
df = pd.read_csv('c3.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


In [76]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [77]:
# Model positives
subset = df [df.actual == 'Defect']
subset

actual     model1     model2     model3
13   Defect  No Defect     Defect     Defect
30   Defect     Defect  No Defect     Defect
65   Defect     Defect     Defect     Defect
70   Defect     Defect     Defect     Defect
74   Defect  No Defect  No Defect     Defect
87   Defect  No Defect     Defect     Defect
118  Defect  No Defect     Defect  No Defect
135  Defect     Defect  No Defect     Defect
140  Defect  No Defect     Defect     Defect
147  Defect     Defect  No Defect     Defect
163  Defect     Defect     Defect     Defect
171  Defect  No Defect     Defect     Defect
176  Defect  No Defect     Defect     Defect
186  Defect  No Defect  No Defect  No Defect
194  Defect     Defect  No Defect     Defect
196  Defect     Defect  No Defect  No Defect

In [78]:
(subset.actual == subset.model1).mean()

0.5

In [79]:
#Model 1 recall
model_recall = (subset.actual == subset.model1).mean()

print(f"Model 1 recall: {model_recall:.2%}")

Model 1 recall: 50.00%


In [80]:
# Model 2 recall
model_recall = (subset.actual == subset.model2).mean()

print(f"Model 2 recall: {model_recall:.2%}")

Model 2 recall: 56.25%


In [81]:
# Model 3 recall
model_recall = (subset.actual == subset.model3).mean()

print(f"Model 3 recall: {model_recall:.2%}")

Model 3 recall: 81.25%


**Takeaways:**

- Quality Control should select a model with higher recall, so model 3 (to avoid false negatives)<br>

In [82]:
subset = df [df.model1 == 'Defect']
subset

actual  model1     model2     model3
3    No Defect  Defect     Defect     Defect
30      Defect  Defect  No Defect     Defect
62   No Defect  Defect  No Defect  No Defect
65      Defect  Defect     Defect     Defect
70      Defect  Defect     Defect     Defect
135     Defect  Defect  No Defect     Defect
147     Defect  Defect  No Defect     Defect
163     Defect  Defect     Defect     Defect
194     Defect  Defect  No Defect     Defect
196     Defect  Defect  No Defect  No Defect

In [83]:
model_precision = (subset.actual == subset.model1).mean()

print(f'Model 1 precision: {model_precision:.2%}')

Model 1 precision: 80.00%


In [84]:
subset = df [df.model2 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model2).mean()

print(f'Model 2 precision: {model_precision:.2%}')

Model 2 precision: 10.00%


In [85]:
subset = df [df.model3 == 'Defect']

model_precision = (subset.actual == subset.model3).mean()

print(f'Model 3 precision: {model_precision:.2%}')

Model 3 precision: 13.13%


**Takeaway for Marketing:**

- Use model 1 since it will minimize the false positive predictions of defects

## 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

 a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?<br>
 b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?<br>
 c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?<br>


In [86]:
df = pd.read_csv('gives_you_paws.csv')
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [87]:
df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [88]:
df['baseline'] = df.actual.value_counts().idxmax()
df

actual model1 model2 model3 model4 baseline
0       cat    cat    dog    cat    dog      dog
1       dog    dog    cat    cat    dog      dog
2       dog    cat    cat    cat    dog      dog
3       dog    dog    dog    cat    dog      dog
4       cat    cat    cat    dog    dog      dog
...     ...    ...    ...    ...    ...      ...
4995    dog    dog    dog    dog    dog      dog
4996    dog    dog    cat    cat    dog      dog
4997    dog    cat    cat    dog    dog      dog
4998    cat    cat    cat    cat    dog      dog
4999    dog    dog    dog    dog    dog      dog

[5000 rows x 6 columns]

In [89]:
(df.actual == df.baseline).mean()

0.6508

In [90]:
model_accuracy = []

for model in df.columns[1:]:
    accuracy = (df.actual == df[model]).mean()
    model_accuracy.append([model, accuracy])

model_accuracy

[['model1', 0.8074],
 ['model2', 0.6304],
 ['model3', 0.5096],
 ['model4', 0.7426],
 ['baseline', 0.6508]]

In [96]:
pd.DataFrame(model_accuracy, columns = ('Model', 'Accuracy'))

Model  Accuracy
0    model1    0.8074
1    model2    0.6304
2    model3    0.5096
3    model4    0.7426
4  baseline    0.6508

**Takeways:** 

 - Model 1 and model 4 perform better than baseline for the accuracy metric.


In [97]:
# team deals with only dog pictures
# dog = positive class
# cat = negative class

In [98]:
model_precision = []

for model in df.columns[1:]:
    
    subset = df [df[model] == 'dog']
    
    precision = (subset.actual == subset[model]).mean()

    model_precision.append([model,precision])
    
model_precision

[['model1', 0.8900238338440586],
 ['model2', 0.8931767337807607],
 ['model3', 0.6598883572567783],
 ['model4', 0.7312485304490948],
 ['baseline', 0.6508]]

In [99]:
pd.DataFrame(model_precision, columns =('Model', 'Precision'))

Model  Precision
0    model1   0.890024
1    model2   0.893177
2    model3   0.659888
3    model4   0.731249
4  baseline   0.650800

**Takeaway for Dog team:**

 - We had to maximize precision to minimize the false positives therefore, we could use model 2 since its the highest with 89.31%

In [101]:
# team deals with only cat pictures
# cat = positive class
# dog = negative class

In [102]:
model_precision = []

for model in df.columns[1:]:
    
    subset = df[df[model] == 'cat']
    
    precision = (subset['actual'] == subset[model]).mean()
    
    model_precision.append([model,precision])
    
model_precision    

[['model1', 0.6897721764420747],
 ['model2', 0.4841220423412204],
 ['model3', 0.358346709470305],
 ['model4', 0.8072289156626506],
 ['baseline', nan]]

In [103]:
pd.DataFrame(model_precision, columns = ('Model', 'Precision'))

Model  Precision
0    model1   0.689772
1    model2   0.484122
2    model3   0.358347
3    model4   0.807229
4  baseline        NaN

**Takeaway for cat team:**

 - We had to maximize precision to minimize the false positives therefore, we could use model 4 since its the highest with 80.72%

## 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

 - sklearn.metrics.accuracy_score
 - sklearn.metrics.precision_score
 - sklearn.metrics.recall_score
 - sklearn.metrics.classification_report

In [112]:
accuracy_score(df.actual, df.model1)

0.8074

In [107]:
print('Model 1')
pd.DataFrame(classification_report(df.actual, df.model1,labels = ['cat', 'dog'], output_dict = True)).T

Model 1


precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [108]:
print('Model 2')
pd.DataFrame(classification_report(df.actual, df.model2,labels = ['cat', 'dog'], output_dict = True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [109]:
print('Model 3')
pd.DataFrame(classification_report(df.actual, df.model3,labels = ['cat', 'dog'], output_dict = True)).T

Model 3


precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [110]:
print('Model 4')
pd.DataFrame(classification_report(df.actual, df.model4,labels = ['cat', 'dog'], output_dict = True)).T

Model 4


precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000